pkg/controller/egress/controller.go

```diff
@@ -228,4 +478,37 @@ func (c *EgressController) deleteEgress(obj interface{}) {
 	c.egressGroupStore.Delete(egress.Name)
 	// Unregister the group from the grouping interface.
 	c.groupingInterface.DeleteGroup(egressGroupType, egress.Name)
+	c.queue.Add(egress.Name)
+}
+
+// addExternalIPPool processes ExternalIPPool ADD events and creates corresponding IP allocator.
+func (c *EgressController) addExternalIPPool(obj interface{}) {
+	pool := obj.(*egressv1alpha2.ExternalIPPool)
+	klog.InfoS("Processing ExternalIPPool ADD event", "pool", pool.Name, "ipRanges", pool.Spec.IPRanges)
+	c.createOrUpdateIPAllocator(pool)
+	c.enqueueEgresses(pool.Name)
+}
+
+// enqueueEgresses enqueues all Egresses that are associated with the provided ExternalIPPool.
+func (c *EgressController) enqueueEgresses(poolName string) {
+	objects, _ := c.egressIndexer.ByIndex(externalIPPoolIndex, poolName)
+	for _, object := range objects {
+		egress := object.(*egressv1alpha2.Egress)
+		c.queue.Add(egress.Name)
+	}
+}
+
+func (c *EgressController) updateExternalIPPool(_, cur interface{}) {
+	pool := cur.(*egressv1alpha2.ExternalIPPool)
+	klog.InfoS("Processing ExternalIPPool UPDATE event", "pool", pool.Name, "ipRanges", pool.Spec.IPRanges)
+	if c.createOrUpdateIPAllocator(pool) {
+		c.enqueueEgresses(pool.Name)
+	}
+}
+
+// deleteEgress processes Egress DELETE events and deletes corresponding EgressGroup.
+func (c *EgressController) deleteExternalIPPool(obj interface{}) {
```

Comment:
Should we enqueue Egresses here to reclaim the Egress IPs?

Reply:
Good point. I was thinking whether we should have validationwebhook for externalIPPool so that one in use cannot be deleted, or just reclaim the IPs after the pool is deleted. Guess the latter one is more friendly to use. I could make the change.

Reply:
Updated the code to reclaim IPs when pool is deleted.

Reply:
I think refusing to delete an externalIPPool that's in use may be a more user-friendly solution actually, as it could help prevent mistakes.

Reply:
I was thinking the scenario that user applying a manifest that has both egresses and externalIPPools. In current implementation, it's ok to create and delete egress and externalIPPool in any order. If we add the check, then externalIPPool can only be deleted after controller processes all egress delete events. But this may be not really important as pool is likely more persistent and your point of preventing mistakes makes sense to me. I'm ok with either strategy. @jianjuns do you have a strong opinion?

Reply:
BTW, I checked Tier resource already has the check that tier in use can not be deleted.

Reply:
I have no strong opinion, but slightly prefer to allow deletion without check. We can have more and more such relations between CRDs; feel we should categorize some as weak references and allow deletion with the references. Pods applied the Egress might lose egress connectivity after the pool is deleted, but seems not very bad to me.
I saw APIs using either strategy.

Reply:
> BTW, I checked Tier resource already has the check that tier in use can not be deleted.

It's true, but we also have a check to prevent a policy from being created in a Tier that doesn't exist. Which we don't want to have for Egress.

I see merits to both approaches.
- Allowing deletion:
  * symmetry with the creation case (one can create an Egress referencing a non-existing pool)
  * may be convenient if the user's intent is indeed to release all IPs at once
- Preventing deletion:
  * prevent errors if the user mistakenly thought that the pool was no longer used
  * symmetry with what we would do for Pod IP pools, as in that case it wouldn't make sense IMO to allow deletion of a pool with IPs in use by Pods

What's the plan for supporting removing IP addresses from an EgressIPPool (e.g. removing an IP address range, which may be in use)? I think we should be consistent across both situations. If we don't plan to have a check for that case, I don't think we should have a check for deletion either.

Reply:
Supporting removing or updating an existing IP address range requires more code to take care of cases. For example, changing 1.1.1.10-1.1.1.20 to 1.1.1.10-1.1.1.30 just means removing the former and adding the latter from code's perspective. We need to first know that the Egress IP's original ipRange is deleted (need to track which ipRange the IP is allocated from then) and then release the IP from the former pool and re-allocate them in the new pool. It can be done but I'm not sure if it's worth. One of my TODOs is to add a validatingwebhook to verify only new range can be added and existing ranges cannot be deleted. Guess to keep consistent experience, we propbably should have a check for deletion too.
@jianjuns @antoninbas how about let's start with simpler solution to get the core feature done first then consider optimizing the user experience, i.e. ipPool can expand and cannot shrink, ipPool can only be removed when it's no longer used.

Reply:
Yes, in my mind we can just support extending the CIDRs, which is also the strategy of many other IPAM solutions I know.
Again, I feel we should finally categorize some relations/references as weak relations, and allow object deletion with a weak relation (and IMO Egress - IPPool relation can be weak).
I am fine to do fix vadliation later.

Reply:
@jianjuns @antoninbas do you have any remaining comments want me to address in this PR? I plan to merge this first to make the regular operation work and will have a separate PR for validation enhancement.